src/librbd/ImageConfigProxy.h

```diff
@@ -0,0 +1,204 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+
+#pragma once
+
+#include <type_traits>
+#include "common/config.h"
+#include "common/config_obs.h"
+#include "common/config_obs_mgr.h"
+#include "common/ceph_mutex.h"
+#include "common/config_proxy.h"
+
+// @c ImageConfigProxy is a different version of ConfigProxy specialised 
+// for per image configuration in order to skip unneeded copies of md_config_t(~2.5x slower).
+namespace librbd {
+class ImageConfigProxy {
+  /**
+   * The current values of all settings described by the schema
+   */
+  ConfigValues values;
+  using md_config_obs_t = ceph::md_config_obs_impl<ImageConfigProxy>;
+  ObserverMgr<md_config_obs_t> obs_mgr;
+  // Reference to the daemon configuration as it includes the schema of daemon related
+  // configurations and per image configuration which will be overriden by "ConfigValues"
+  md_config_t &config; // FIXME: use shared_ptr in case of dangling reference
```

Comment:
`md_config_t` is stored in `ConfigProxy`, which is stored in `CephContext`. `CephContext` itself is reference-counted, so you can avoid using `shared_ptr` here if `ImageCtx` holds a reference to `CephContext`:
```diff
-    CephContext *cct;
+    boost::intrusive_ptr<CephContext> cct;
```

Reply:
Is it preferrable to md_config_t? That means g_conf() will be called inside the ImageConfigProxy or the constructor will need a CephContext which looks more convoluted than necessary.

Reply:
i wasn't proposing any change to `ImageConfigProxy`; just to the existing `ImageCtx::cct` variable

by holding a ref count on `cct` there, you can guarantee the lifetime of this `md_config_t &config` reference

Reply:
Done!